src/main/java/com/nirima/jenkins/plugins/docker/DockerContainerWatchdog.java

```diff
@@ -0,0 +1,538 @@
+             * he manually force-removed the containers from the docker instance). 
+             * 
+             * At the end, this means that we cannot verify strictly that the containerId stored with this
+             * node, really exists on (any of the) DockerCloud instances or not. 
+             * The only option is that, as long as we see "some container having an id like this" on "some of
+             * our DockerCloud instances", then we should be careful with it, and do not delete the node (yet).
+             * Even if the containerId originates from a collision, then we would just postpone the deletion
+             * of the node up to that point in time, when that container gets deleted from the DockerContainer
+             * instance (on the next recurrence of this watchdog, we then would detect that situation and the
+             * node gets deleted automatically, as no container with the appropriate identifier could be detected
+             * anymore). 
+             * 
+             * It is to be hoped that collision on container identifiers are happening rarely enough to such that
+             * we do not run into troubles here. Otherwise, still a larger set of broken nodes could pile up...
+             */
+            
+            boolean seenOnDockerInstance = csmMerged.isContainerIdRegistered(dtn.getContainerId());
+            if (seenOnDockerInstance) {
+                // the node and the container still might have a proper mapping => ignore
+                continue;
+            }
+            
+            SlaveComputer computer = dtn.getComputer();
+            if (computer == null) {
+                // Probably the node is being closed down right now, so we shouldn't touch it.
+                continue;
+            }
+            
+            if (!computer.isOffline()) {
+                // the node is still running; we should not touch it.
+                continue;
+            }
+            
+            // the container is already gone for the node, but the node did not notice it yet properly
+            LOGGER.info("{} has container ID {}, but the container does not exist in any docker cloud. Will remove node.", dtn, dtn.getContainerId());
+
+            try {
+                removeNode(dtn);
+                executionStatistics.nodesRemoved++;
+            } catch (IOException e) {
+                LOGGER.warn(String.format("Failed to remove orphaned %s.", dtn), e);
+                executionStatistics.nodesRemovedFailed++;
+            }
+        }
+    }
+
+    private void checkForTimeout(Instant startedTimestamp) {
+        final Instant now = clock.instant();
+        Duration runtime = Duration.between(startedTimestamp, now);
+        
+        if (runtime.compareTo(PROCESSING_TIMEOUT_IN_MS) > 0) {
+            throw new WatchdogProcessingTimeout();
+        }
+    }
+    
+    private static class Statistics {
+        public long executions;
```

Comment:
Do these all need to be `public`?  Won't package-level access be sufficient?
Do they all need to be `long`?  Is an `int` too small for all of them?

Reply:
ad `public`: No, but it theoretically also does not hurt, as the class is set to `private`. 

ad `long` vs. `int`: Some of them can become very large (e.g. all the `.*RuntimeSum`s). Remember that our plugin may have a lifetime in the order of months. If we count milliseconds there, this can become quite a bit.
I don't think that we will have `long`-ish number of executions, though. Yet, I did not want to look the statistics to be typed differently: We then also would get to a very strange discussion about what we may expect and what not. 
Moreover, the class is a singleton anyway, so we are talking about roughly `4*10` bytes of memory consumption here -- at best.

If you insist to save that memory, then we may discuss, if we want to make the "counters of events" to only be `int`s.